code/modules/mod/mod_modules.dm

```diff
@@ -0,0 +1,616 @@
+			UnregisterSignal(mod.wearer, COMSIG_MOB_MIDDLECLICKON)
+			UnregisterSignal(mod.wearer, COMSIG_MOB_ALTCLICKON)
+	return TRUE
+
+/// Called when the module is used
+/obj/item/mod/module/proc/on_use()
+	if(!COOLDOWN_FINISHED(src, cooldown_timer))
+		return FALSE
+	if(!drain_power(use_power_cost))
+		return FALSE
+	COOLDOWN_START(src, cooldown_timer, cooldown_time)
+	return TRUE
+
+/// Called when an activated module without a device is used with middle/alt click
+/obj/item/mod/module/proc/on_select_use(mob/source, atom/target)
+	SIGNAL_HANDLER
+
+	if(!on_use())
+		return NONE
+	return COMSIG_MOB_CANCEL_CLICKON
+
+/// Called on the MODsuit's process
+/obj/item/mod/module/proc/on_process(delta_time)
+	if(active)
+		if(!drain_power(active_power_cost * delta_time))
+			on_deactivation()
+			return FALSE
+	else
+		drain_power(idle_power_cost * delta_time)
+	return TRUE
+
+/// Drains power from the suit cell
+/obj/item/mod/module/proc/drain_power(amount)
+	if(!mod.cell || (mod.cell.charge < amount))
+		return FALSE
+	mod.cell.charge = max(0, mod.cell.charge - amount)
+	return TRUE
+
+/// Adds additional things to the MODsuit ui_data()
+/obj/item/mod/module/proc/add_ui_data()
+	return list()
+
+/// Called when the device moves to a different place on active modules
+/obj/item/mod/module/proc/on_exit(datum/source, atom/movable/offender, atom/newloc)
+	SIGNAL_HANDLER
+
+	if(newloc == mod.wearer || newloc == src)
+		return
+	if(offender == device)
+		on_deactivation()
+
+/// Called when the device gets deleted on active modules
+/obj/item/mod/module/proc/on_device_deletion(datum/source)
+	SIGNAL_HANDLER
+
+	if(source == device)
+		device = null
+		qdel(src)
+
+/obj/item/mod/module/storage
+	name = "MOD storage module"
+	desc = "A module using nanotechnology to fit a storage inside of the MOD."
+	complexity = 5
+	incompatible_modules = list(/obj/item/mod/module/storage)
+	var/datum/component/storage/concrete/storage
+	var/max_w_class = WEIGHT_CLASS_SMALL
+	var/max_combined_w_class = 14
+	var/max_items = 7
+
+/obj/item/mod/module/storage/antag
+	name = "MOD syndicate storage module"
+	max_w_class = WEIGHT_CLASS_NORMAL
+	max_combined_w_class = 21
+	max_items = 21
+
+/obj/item/mod/module/storage/antag/wiz
+	name = "MOD enchanted storage module"
+
+/obj/item/mod/module/storage/Initialize()
+	. = ..()
+	storage = AddComponent(/datum/component/storage/concrete)
+	storage.max_w_class = max_w_class
+	storage.max_combined_w_class = max_combined_w_class
+	storage.max_items = max_items
+
+/obj/item/mod/module/storage/on_install()
+	var/datum/component/storage/modstorage = mod.AddComponent(/datum/component/storage, storage)
+	modstorage.max_w_class = max_w_class
+	modstorage.max_combined_w_class = max_combined_w_class
+	modstorage.max_items = max_items
+
+/obj/item/mod/module/storage/on_uninstall()
+	var/datum/component/storage/modstorage = mod.GetComponent(/datum/component/storage)
+	modstorage.RemoveComponent()
+
+/obj/item/mod/module/visor
```

Comment:
remember to give nuke ops something that gives them flash protection, if you haven't done that already